### PR TITLE
Undef aware equality

### PR DESCRIPTION
**Pull request by leonerd**
Currently this draft PR contains two alternative suggestions, as raised by various discussions on perl5-porters@. We can continue to discuss the relative merits here as a draft, until we eventually decide to reject at least one of them, at which point the other will stand as a proposal to be submitted.

### CODE REVIEW

**Comment by exodist** [comment]
I am 100% in favor of either one. I think the one that lets you use :u and :i is ideal. But I would be satisfied with the other one too.

**Comment by guest20** [comment]
Feels like it should be a `/`, like regex modifiers ... but I guess that's taken by regex modifiers.

**Comment by leonerd** [comment]
> Feels like it should be a `/`, like regex modifiers ... but I guess that's taken by regex modifiers.

There is an entire section in the document under "Rejected Ideas" explaining why it can't.

File: ppcs/ppcTODO-metaoperator-flags.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+## Rationale
+
+(explain why the (following) proposed solution will solve it)
+
+## Specification
+
+Following the name of a regular value-returning infix operator, an optional set of flag letters can be supplied after a separating colon:
+
+```perl
+$x eq:u $y
```
Comment (rabbiveesh) [inline ppcs/ppcTODO-metaoperator-flags.md:29]: I added prospective support for highlighting this to tree-sitter-perl; looks pretty solid, actually
image

**Comment by Grinnz** [comment]
I think either option is nice, but I slightly favor the "modifiers" version:

If nothing else, it's a slightly different spelling of the other version that makes it clearer what's going on: eq:u vs equ is more obviously an eq operator.

If someday we decide the semantics are non-ideal we can fix them by just adding another modifier rather than a whole new operator, though this wouldn't be a huge problem other than spelling.

And least importantly, we can add other modifiers. not very important because there aren't any others that are obviously useful at this time. But maybe with the option available, the need will present itself.

**Comment by Grinnz** [comment]
As a practical matter to what semantics we might change our mind on, I could go either way on whether these operators should consider undef equal to itself.

**Comment by xiaoyafeng** [comment]
I  personally prefer modifiers one. Defined or not is a important concept in Perl,  But perl lack a neat and consistent way to deal with it. perl already has defined, undef, // and maybe ppc0021. but what all does just makes "defined or not"  more scattered in perl.   I think modifier is a good way to solve this problem.

**Comment by guest20** [comment]
@leonerd do the debates still stay in github issues?

**Comment by leonerd** [comment]
> @leonerd do the debates still stay in github issues?

That's a wider process question - perhaps @book @haarg @ap as the current PSC could comment on that.

**Pull request merged**